types/underscore/underscore-tests.ts

```diff
@@ -647,8 +663,184 @@ interface ChainTypeExtractor {
+
+    const anyPropertyNameteratee: _.Iteratee<any, string> = stringRecordProperty;
+    anyPropertyNameteratee; // $ExpectType string
+
+    // property paths
+    const listPropertyPathIteratee: _.Iteratee<_.List<StringRecord>, string> = stringRecordPropertyPath;
+    listPropertyPathIteratee; // $ExpectType (string | number)[]
+
+    const dictionaryPropertyPathIteratee: _.Iteratee<_.Dictionary<StringRecord>, string> = stringRecordPropertyPath;
+    dictionaryPropertyPathIteratee; // $ExpectType (string | number)[]
+
+    const unionCollectionItemPropertyPathIteratee: _.Iteratee<_.List<IntersectingProperties>, string | boolean> = stringRecordPropertyPath;
+    unionCollectionItemPropertyPathIteratee; // $ExpectType (string | number)[]
+
+    const unionCollectionPropertyPathIteratee: _.Iteratee<StringRecord[] | _.Dictionary<StringRecord>, string> = stringRecordPropertyPath;
+    unionCollectionPropertyPathIteratee; // $ExpectType (string | number)[]
+
+    const anyPropertyPathIteratee: _.Iteratee<any, string> = stringRecordPropertyPath;
+    if (_.isArray(anyPropertyPathIteratee)) {
+        anyPropertyPathIteratee; // $ExpectType (string | number)[]
+    }
+
+    // identity
+    const listIdentityIteratee: _.Iteratee<_.List<StringRecord>, string> = undefined;
+    listIdentityIteratee; // $ExpectType undefined
+
+    const dictionaryIdentityIteratee: _.Iteratee<_.Dictionary<StringRecord>, string> = undefined;
+    dictionaryIdentityIteratee; // $ExpectType undefined
+
+    const unionCollectionItemIdentityIteratee: _.Iteratee<_.List<IntersectingProperties>, string | boolean> = undefined;
+    unionCollectionItemIdentityIteratee; // $ExpectType undefined
+
+    const unionCollectionIdentityIteratee: _.Iteratee<StringRecord[] | _.Dictionary<StringRecord>, string> = undefined;
+    unionCollectionIdentityIteratee; // $ExpectType undefined
+
+    const anyIdentityIteratee: _.Iteratee<any, string> = undefined;
+    anyIdentityIteratee; // $ExpectType undefined
+}
+
+// IterateeResult
+declare const functionIterateeResult: _.IterateeResult<() => string, StringRecord>;
+functionIterateeResult; // $ExpectType string
+
+declare const partialObjectIterateeResult: _.IterateeResult<Partial<StringRecord>, StringRecord>;
+partialObjectIterateeResult; // $ExpectType boolean
+
+declare const knownPropertyNameIterateeResult: _.IterateeResult<typeof stringRecordProperty, IntersectingProperties>;
+knownPropertyNameIterateeResult; // $ExpectType string | boolean
+
+declare const unknownPropertyNameIterateeResult: _.IterateeResult<typeof stringRecordProperty, NonIntersectingProperties>;
+unknownPropertyNameIterateeResult; // $ExpectType any
+
+declare const propertyPathIterateeResult: _.IterateeResult<_.EnumerableKey[], StringRecord>;
+propertyPathIterateeResult; // $ExpectType any
+
+declare const identityIterateeResult: _.IterateeResult<undefined, StringRecord>;
+identityIterateeResult; // $ExpectType StringRecord
```

Comment:
Would it be worth dropping the `undefined` cases from `Iteratee` and `IterateeResult`? The only function I expect to use `IterateeResult` is `map`, and we decided to make `Iteratee` non-optional there, but allowing `undefined` in `Iteratee` undermines that a bit. The optionalness of the `iteratee` parameter via `?` covers undefined in other cases, so I feel like it might not be necessary to include it.

Reply:
No, we shouldn't, because someone might want to be able to pass an iteratee that might or might not be undefined.

Reply:
Yes, or even explicitly pass `undefined` or `null`.

By the way, I realized that the current implementation of `IterateeResult` is still incomplete. I wrote `extends void` in the first case, but that should be `extends void | null`. I assumed that was equivalent, but it isn't under the assumption of `strictNullChecks`.

Reply:
Good point, I'll update that, thanks! I think I'll move them to the ends of these types though since I feel like they are less likely to be used as compared to other options.

Reply:
It is the least expensive check, though.

Reply:
You're not wrong, but even a small expense can add up when it's always evaluated before other possibilities that I personally at least feel are much much more likely.

That being said, I don't feel strongly about this, so if you would much prefer to see this moved up to the beginning of these types again I'll be happy to oblige.